### PR TITLE
feat: Generic text flow

### PR DESCRIPTION
**Pull request by jmars**
image

### CODE REVIEW

File: packages/article-image/src/article-image.base.js [hunk 1]
```diff
@@ -54,40 +60,41 @@ const ArticleImage = ({ imageOptions, captionOptions }) => {
...
-  const renderedCaption = (
-    <Context.Consumer>
-      {({ theme }) => renderCaption(caption, credits, display, theme)}
-    </Context.Consumer>
-  );
+  const captionChildren = [renderCaption(caption, credits, display, theme)];
 
-  if (!display || !ratio) {
-    return renderedCaption;
+  if (!ratio) {
```
Comment (jmars) [inline packages/article-image/src/article-image.base.js]: Should also check display

File: packages/article-image/src/article-image.base.js [hunk 2]
```diff
@@ -54,40 +60,41 @@ const ArticleImage = ({ imageOptions, captionOptions }) => {
...
+    ...captionChildren
+  ];
 };
 
 ArticleImage.propTypes = propTypes;
 ArticleImage.defaultProps = defaultPropTypes;
 
-export default ArticleImage;
+export default props => (
+  <Context.Consumer>
```
Comment (jmars) [inline packages/article-image/src/article-image.base.js]: ?

File: packages/article-image/src/inline-image.js [hunk 3]
```diff
@@ -17,21 +22,70 @@ const renderCaption = (display, caption, credits) => {
...
+              </View>
+            )}
+          </InlineElement>
+        </FlowText>
+      </View>
+    );
+  }
   return [
-    <View key="img" style={styles.inlineImage}>
+    <View key="img">
```
Comment (jmars) [inline packages/article-image/src/inline-image.js]: style for mobile?

File: packages/article-image/src/inline-image.js [hunk 4]
```diff
@@ -17,21 +22,70 @@ const renderCaption = (display, caption, credits) => {
...
+          style={{ maxWidth }}
+          textStyle={styles.articleTextElement}
+        >
+          <InlineElement align="left" start={0}>
+            {style => (
+              <View
+                key="img"
+                style={[
+                  styles.inlineImage,
+                  { height: style.height, maxWidth: "35%", width: style.width }
```
Comment (jmars) [inline packages/article-image/src/inline-image.js]: Should be a stylesheet?

File: packages/article-image/src/inline-image.js [hunk 5]
```diff
@@ -17,21 +22,70 @@ const renderCaption = (display, caption, credits) => {
...
+      >
+        <FlowText
+          localRender={localRender}
+          markup={children}
+          style={{ maxWidth }}
+          textStyle={styles.articleTextElement}
+        >
+          <InlineElement align="left" start={0}>
+            {style => (
+              <View
```
Comment (jmars) [inline packages/article-image/src/inline-image.js]: Same as below can be a function

File: packages/article-image/src/inline-image.js [hunk 6]
```diff
@@ -40,11 +94,24 @@ const InlineImage = ({ imageOptions, captionOptions }) => {
...
 };
 
-InlineImage.propTypes = propTypes;
-InlineImage.defaultProps = defaultPropTypes;
+FlowImage.propTypes = propTypes;
+FlowImage.defaultProps = defaultPropTypes;
 
-export default InlineImage;
+export default props => (
+  <Context.Consumer>
```
Comment (jmars) [inline packages/article-image/src/inline-image.js]: should be passed from article-body-row

File: packages/article-image/src/styles/index.js [hunk 7]
```diff
@@ -11,68 +9,87 @@ const containerShared = {
...
-  secondaryImage: {
-    width: "50%"
-  }
-});
+export default (scale, isTablet) => {
+  const { colours, fontFactory, spacing } = styleguide({ scale });
 
-export default styles;
+  return {
+    articleMainContentRow: {
```
Comment (jmars) [inline packages/article-image/src/styles/index.js]: Duplicated styles

File: packages/article-magazine-comment/__tests__/web/__snapshots__/article-with-style.web.test.js.snap [hunk 8]
```diff
@@ -489,7 +489,7 @@ exports[`full article with style 1`] = `
   font-size: 17;
   line-height: 26;
   color: #333333;
-  margin-bottom: 25px;
+  margin-bottom: 10px;
```
Comment (jmars) [inline packages/article-magazine-comment/__tests__/web/__snapshots__/article-with-style.web.test.js.snap]: margins on web?

File: packages/article-magazine-comment/__tests__/web/__snapshots__/semantic.web.test.js.snap [hunk 9]
```diff
@@ -74,9 +74,7 @@ exports[`1. a full article 1`] = `
     url="https://twitter.com/@AName"
   >
     “
-    <cite>
-      The pull quote content
-    </cite>
+
```
Comment (jmars) [inline packages/article-magazine-comment/__tests__/web/__snapshots__/semantic.web.test.js.snap]: missing content

File: packages/article-paragraph/__tests__/android/__snapshots__/article-paragraph-with-style.android.test.js.snap [hunk 10]
```diff
@@ -6,48 +6,61 @@ exports[`1. paragraph with a drop cap 1`] = `
...
-      Object {
-        "alignItems": "flex-start",
-        "flex": 0,
-        "flexDirection": "row",
-        "height": 1,
-        "justifyContent": "flex-start",
-      }
-    }
-  >
+  <View>
```
Comment (jmars) [inline packages/article-paragraph/__tests__/android/__snapshots__/article-paragraph-with-style.android.test.js.snap]: Why is this object gone now

File: packages/article-paragraph/__tests__/renderer.js [hunk 11]
```diff
@@ -8,6 +8,24 @@ import ArticleParagraph from "../src";
...
+  const renderers = {
+    ...coreRenderers,
+    paragraph(key, attributes, children, indx, node) {
+      return {
+        element: (
+          <ArticleParagraph
+            ast={node}
+            dropCapFont={theme.dropCapFont}
+            key={indx}
+            localRender={renderers}
```
Comment (jmars) [inline packages/article-paragraph/__tests__/renderer.js]: Lift inlines up into article-body-row

File: packages/article-paragraph/src/index.js [hunk 12]
```diff
@@ -24,20 +25,29 @@ const ArticleParagraphWrapper = ({
   const { name, attributes } = astChildren[0];
   if (name === "dropCap") {
     const { value } = attributes;
-    const text = renderTreeAsText(ast).slice(1);
     return (
       <DropCapWrapper
```
Comment (jmars) [inline packages/article-paragraph/src/index.js]: paragraph shouldnt know about dropcaps

File: packages/article-skeleton/__tests__/android/__snapshots__/article-header-with-style.android.test.js.snap [hunk 13]
```diff
@@ -143,6 +143,28 @@ exports[`1. article with header 1`] = `
               }
             }
             dropCapColour="#FF0000"
+            localRender={
```
Comment (jmars) [inline packages/article-skeleton/__tests__/android/__snapshots__/article-header-with-style.android.test.js.snap]: Filter from snapshots

File: packages/article-skeleton/__tests__/android/__snapshots__/article-with-style.android.test.js.snap [hunk 14]
```diff
@@ -192,9 +192,7 @@ exports[`1. a full article with all content items with dropcap template 1`] = `
           }
         >
           <View>
-            <PullQuote>
-              The pull quote content
-            </PullQuote>
+            <PullQuote />
```
Comment (jmars) [inline packages/article-skeleton/__tests__/android/__snapshots__/article-with-style.android.test.js.snap]: Investigate

File: packages/article-skeleton/src/article-skeleton.js [hunk 15]
```diff
@@ -144,13 +145,40 @@ class ArticleSkeleton extends Component {
...
 
-    const newContent = [...dataSource.content];
-    if (newContent && newContent.length > 0) {
-      newContent[0] = insertDropcapIntoAST(
-        newContent[0],
-        template,
-        dropcapsDisabled
-      );
+    let newContent = [...dataSource.content];
+    for (let i = 0; i < newContent.length; i += 1) {
```
Comment (jmars) [inline packages/article-skeleton/src/article-skeleton.js]: should be a function (with unit tests)

File: packages/markup/src/flatten.js [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+  link(key, attr, renderedChildren, indx, node) {
+    return {
+      element: flattenChildren(renderedChildren).map(child => ({
+        ...child,
+        markup: [node, ...child.markup]
+      })),
+      shouldRenderChildren: true
+    };
+  },
+  paragraph(key, attr, renderedChildren, indx, node) {
```
Comment (jmars) [inline packages/markup/src/flatten.js]: lots of duplicated flattens here

File: packages/pull-quote/__tests__/android/__snapshots__/pull-quotes-with-style.android.test.js.snap [hunk 17]
```diff
@@ -73,35 +72,12 @@ exports[`different colours 1`] = `
...
-          "borderLeftWidth": 1,
-          "display": "flex",
-          "flexDirection": "row",
-          "height": 15,
-          "marginLeft": 7,
-          "paddingLeft": 7,
-        }
-      }
-    >
+    <View>
```
Comment (jmars) [inline packages/pull-quote/__tests__/android/__snapshots__/pull-quotes-with-style.android.test.js.snap]: cant see children again in snapshot

File: packages/pull-quote/__tests__/shared-with-style.base.js [hunk 18]
```diff
@@ -12,6 +13,7 @@ export default renderComponent => {
         caption={caption}
         onTwitterLinkPress={() => null}
         quoteColour="#850029"
+        renderedChildren={["Some content"]}
```
Comment (jmars) [inline packages/pull-quote/__tests__/shared-with-style.base.js]: ?

File: packages/pull-quote/__tests__/web/__snapshots__/themes-with-style.web.test.js.snap [hunk 19]
```diff
@@ -109,10 +109,10 @@ exports[`1. pull quote in culture magazine 1`] = `
     </div>
     <div
       className="c0 S2"
+      isTablet={false}
+      scale="medium"
     >
-      <cite>
-        Some content
-      </cite>
+      Some content
```
Comment (jmars) [inline packages/pull-quote/__tests__/web/__snapshots__/themes-with-style.web.test.js.snap]: Should be wrapped in a <cite> tag

File: packages/pull-quote/src/pull-quote.base.js [hunk 20]
```diff
@@ -1,37 +1,105 @@
...
+  localRender,
+  twitter,
+  isTablet,
+  scale
+}) => {
+  const maxWidth = screenWidth(isTablet);
+  const styles = styleFactory(scale, isTablet);
+  const content =
+    renderedChildren && renderedChildren.length ? renderedChildren[0] : "";
+  if (isTablet) {
```
Comment (jmars) [inline packages/pull-quote/src/pull-quote.base.js]: does the same thing as dropcaps/images

File: packages/pull-quote/src/pull-quote.base.js [hunk 21]
```diff
@@ -1,37 +1,105 @@
...
-);
+  );
+};
 
 PullQuotes.propTypes = propTypes;
 PullQuotes.defaultProps = defaultProps;
 
-export default PullQuotes;
+export default props => (
+  <Context.Consumer>
```
Comment (jmars) [inline packages/pull-quote/src/pull-quote.base.js]: NO

File: packages/text-flow/src/styles/index.web.js [hunk 22]
```diff
@@ -0,0 +1,12 @@
+import { StyleSheet } from "react-native";
+import sharedStyles from "./shared";
+
+const styles = StyleSheet.create({
+  ...sharedStyles,
+  TextFlowBody: {
+    ...sharedStyles.TextFlowBody,
+    color: "blue"
```
Comment (jmars) [inline packages/text-flow/src/styles/index.web.js]: What is this

File: packages/text-flow/src/wrap-text.js [hunk 23]
```diff
@@ -0,0 +1,139 @@
+/* eslint-disable no-continue,react/no-did-update-set-state,react/forbid-prop-types */
+import React, { Component } from "react";
+import { View, Text } from "react-native";
+import { renderTree } from "@times-components/markup-forest";
+import { flatten } from "@times-components/markup";
+import PropTypes from "prop-types";
+import { measureElements, layoutText, measureContainer } from "./text-flow";
+
+class FlowText extends Component {
```
Comment (jmars) [inline packages/text-flow/src/wrap-text.js]: Should not know about children and should return children unchanged if we are on a mobile

File: packages/article-image/src/styles/index.js [hunk 24]
```diff
@@ -11,68 +9,87 @@ const containerShared = {
...
+      flexDirection: "row",
+      flexWrap: "nowrap",
+      paddingBottom: spacing(5),
+      paddingLeft: spacing(2),
+      paddingRight: spacing(2),
+      width: "100%"
+    },
+    inlineContainerTablet: {
+      ...containerShared,
+      paddingLeft: spacing(0),
```
Comment (aronshamash) [inline packages/article-image/src/styles/index.js]: just `0` is fine

File: packages/article-image/src/styles/index.js [hunk 25]
```diff
@@ -11,68 +9,87 @@ const containerShared = {
...
+    secondaryCaption: {
+      paddingLeft: spacing(2),
+      paddingTop: 0,
+      width: "50%"
+    },
+    secondaryContainer: {
+      flexDirection: "row",
+      flexWrap: "nowrap",
+      paddingBottom: spacing(4),
+      paddingLeft: spacing(2),
```
Comment (aronshamash) [inline packages/article-image/src/styles/index.js]: `paddingHorizontal`

File: packages/markup/src/flatten.js [hunk 26]
```diff
@@ -0,0 +1,64 @@
+const flattenChildren = childrenList =>
+  childrenList.reduce((acc, children) => [...acc, ...children], []);
+
+const flatten = {
+  bold(key, attr, renderedChildren, indx, node) {
+    return {
+      element: flattenChildren(renderedChildren).map(child => ({
```
Comment (aronshamash) [inline packages/markup/src/flatten.js]: create function and reuse

File: packages/pull-quote/src/styles/index.js [hunk 27]
```diff
@@ -1,30 +1,30 @@
...
+    // iOS Font hack
+    caption: {
+      ...shared.caption,
+      paddingTop: 4
+    },
+    container: {
+      ...shared.container,
+      marginBottom: 0,
+      paddingBottom: spacing(2),
+      paddingLeft: spacing(2),
```
Comment (aronshamash) [inline packages/pull-quote/src/styles/index.js]: `paddingHorizontal`

**Pull request closed without merging**